partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/DynamicTyper.kt

```diff
@@ -0,0 +1,323 @@
+import org.partiql.value.PartiQLValueType.STRING
+import org.partiql.value.PartiQLValueType.STRUCT
+import org.partiql.value.PartiQLValueType.SYMBOL
+import org.partiql.value.PartiQLValueType.TIME
+import org.partiql.value.PartiQLValueType.TIMESTAMP
+
+/**
+ * Graph of super types for quick lookup because we don't have a tree.
+ */
+internal typealias SuperGraph = Array<Array<PartiQLValueType?>>
+
+/**
+ * For lack of a better name, this is the "dynamic typer" which implements the typing rules of SQL-99 9.3.
+ *
+ * SQL-99 9.3 Data types of results of aggregations (<case-when>, <collection value expression>, <query expression>)
+ *  > https://web.cecs.pdx.edu/~len/sql1999.pdf#page=359
+ *
+ * Usage,
+ *  To calculate the type of an "aggregation" create a new instance and "accumulate" each possible type.
+ *  This is a pain with StaticType...
+ */
+@OptIn(PartiQLValueExperimental::class)
+internal class DynamicTyper {
+
+    private var supertype = NULL
+    private var args = mutableListOf<PartiQLValueType>()
+
+    private var nullable = false
+    private var missable = false
+    private val types = mutableSetOf<StaticType>()
+
+    fun accumulate(type: StaticType) {
+        val arg = type.toRuntimeType()
+        args.add(arg)
+        if (type is SingleType) {
+            when (type) {
+                is MissingType -> missable = true
+                is NullType -> nullable = true
+                else -> calculate(type, arg)
+            }
+        } else {
+            for (t in type.flatten().allTypes) {
+                when (t) {
+                    is MissingType -> missable = true
+                    is NullType -> nullable = true
+                    else -> calculate(t)
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns a pair of the return StaticType and the coercion
+     *
+     * @return
+     */
+    fun mapping(): Pair<StaticType, List<Pair<PartiQLValueType, PartiQLValueType>>?> {
+        val modifiers = mutableSetOf<StaticType>()
+        if (nullable) modifiers.add(StaticType.NULL)
+        if (missable) modifiers.add(StaticType.MISSING)
+        // If at top supertype, then return union of all accumulated types
+        if (supertype == ANY) {
+            return StaticType.unionOf(types + modifiers) to null
+        }
+        // If a collection, then return union of all accumulated types as these coercion rules are not defined by SQL.
+        if (supertype == STRUCT || supertype == BAG || supertype == LIST || supertype == SEXP) {
+            return StaticType.unionOf(types + modifiers) to null
+        }
+        // Otherwise, return the supertype along with the coercion mapping
+        val type = supertype.toNonNullStaticType()
+        val mapping = args.map { it to supertype }
+        return if (modifiers.isEmpty()) {
+            type to mapping
+        } else {
+            StaticType.unionOf(setOf(type) + modifiers).flatten() to mapping
+        }
```

Comment:
If one would keep accumulating `MISSING`, does not seem like the NULL type will be dropped.

Reply:
Could you please elaborate more?

Reply:
```
val typer = DynamicTyper() // at this stage, super Type is NULL
typer.accumulate(MISSING) // no touch on super type, missable is true

// modifier contains missing, 
// type is supertype.toNonNullStaticType() -> NULL
// return is StaticType.unionOf(setOf(type) + modifiers).flatten() to mapping
typer.mapping() // UnionOf( NULL, MISSING) 
```